qi-doc/scribblings/field-guide.scrbl

```diff
@@ -425,6 +427,29 @@ If you'd like to use Racket's order of effects in any flow, @seclink["Using_Rack
 
 See @secref["Effect_Locality"] for more insights into Qi's handling of effects.
 
+@subsubsection{Schrodinger's Probe}
+
+Another curious thing to watch out for is that use of the @seclink["Using_a_Probe"]{probe debugger} can affect the @seclink["Order_of_Effects"]{order of effects}, as it could suppress optimizations that would be otherwise be performed if the @tech{flow} were unobserved.
+
+Consider this example:
+
+@racketblock[
+(define-flow foo
+  (~> (pass (effect E₁ odd?))) readout (>< (effect E₂ sqr)))
+
+(probe (foo 1 2 3))
+]
+
+Here, with the @racket[readout], all the effects E₁ would occur first, followed by all of the E₂ effects. Without the @racket[readout], the flow would be deforested by the compiler to:
+
+@racketblock[
+  (>< (if (effect E₁ odd?) (effect E₂ sqr) ⏚))
+]
+
+… and the effects E₁ and E₂ would be interleaved. Either order is consistent with @seclink["Effect_Locality"]{locality} but they are @emph{different}. Indeed, the @racket[readout] in this case would not even represent a valid point in the a priori optimized program.
+
+So it's important to bear in mind that one cannot observe a flow using @racket[probe] without changing the program being observed, a change which in some cases has no observable impact, and which in other cases is significant. But now that you understand this phenomenon, you can develop intuition for the nature of such changes, and how best to use the tool to find the answers you are looking for.
```

Comment:
This is a very interesting point and I hope we'll explore it in more depth over time as I suspect Qi could provide a natural setting for understanding this (if it isn't already well understood in the field, which it well might be!).

First re: probe and readout, `readout` is a simple lambda (and thus a flow) that escapes into a continuation bound to a parameter (the parameter is used so that it is in scope everywhere during evaluation). If `probe` is omitted then readout escapes into the continuation that the parameter was defined with, which just raises an error saying that probe wasn't used.

The only thing `probe` does is take a fresh continuation and wrap the flow with it, leaving it otherwise intact.

I don't know what to consider escaping into a continuation, from the perspective of effects. By the loose definition we have in the Field Guide, "an effect is anything a function does that isn't reflected in its inputs and outputs." Escaping prior to returning almost seems to match this definition, except that doing so also seems to make the outputs _undefined_ in the original context of evaluation. So, I don't know if it's an effect or something else, that we could perhaps have a pretty interesting formal model for, which we could maybe think about in a clean functional way.

I am basically making all this up, but this is just to say, I think it will be interesting to explore! I don't know if, for now, we have enough to go on to simplify the explanation of the "Schrodinger's probe" phenomenon, though :)

re: the compiler reasoning across the readout barrier, the issue is that due to optimizations, sometimes `readouts` can be placed in source programs in places that don't represent valid points in the target program, so technically, the values read out at that point in the source program would never be encountered in the target program (e.g. an intermediate list that would not be constructed in a deforested sequence). Yet, conceptually those values do exist just in terms of the logical structure of the program, and as you said, in the absence of effects, whether the values are defined there or not in the running program are irrelevant to the actual output. It's all a bit confusing! I agree we can probably make this clearer.

This discussion may be of interest to @jairtrejo and @dzoep .

Reply:
There's no question of being a stick in the mud! Your commitment to a high standard for the docs is always appreciated.

But I think I agree with your earlier assessment that the effects stuff is potentially quite involved, and best tackled separately from the more surface level stuff that's in this PR. I will aim to separate that out into a new PR soon so we can continue discussions there while unblocking the other stuff that's in here.

Reply:
Well, since you've forced me to come up with clearer explanations 😜 , I've started a new PR dedicated to developing Qi's "theory of effects" ( #165 ). It captures discussions on the topic so far and also starts to develop it more precisely. In thinking about it more, I felt that what we've been calling "effect locality" is actually (at least) two distinct notions -- what the PR now calls "effect locality" and "well-ordering of effects." I also revised the explanation of Schrodinger's probe to hopefully make it clearer but I'm not sure if I've captured your feedback.

For anyone following this discussion, please follow #165 (not sure if there is a way to do that without explicitly commenting on it - but if you thumbs-up this comment I will tag you there. I will probably tag you there in any case if your input is needed, but there is no particular time pressure on this new PR and we can take the time we need so it's reasonably accurate).